source/common/router/upstream_request.cc

```diff
@@ -41,98 +41,108 @@
 namespace Envoy {
 namespace Router {
 
+void UpstreamCodecFilter::onBelowWriteBufferLowWatermark() {
+  callbacks_->clusterInfo()->stats().upstream_flow_control_resumed_reading_total_.inc();
+  callbacks_->upstreamCallbacks()->upstream()->readDisable(false);
+}
+
+void UpstreamCodecFilter::onAboveWriteBufferHighWatermark() {
+  callbacks_->clusterInfo()->stats().upstream_flow_control_paused_reading_total_.inc();
+  callbacks_->upstreamCallbacks()->upstream()->readDisable(true);
+}
+
+void UpstreamCodecFilter::onUpstreamConnectionEstablished() {
+  if (latched_end_stream_.has_value()) {
```

Comment:
Would you expect this to get multiple times? Does the lifetime of the upstream filters live between multiple attempts? If not expected, should this ASSERT?

Reply:
(currently) it can only be called once but shouldn't assert because it's totally possible to have the upstream connection established before the full stream is proxied.